### PR TITLE
Revert the look up and service call when trying with ART and FRT in s…

### PR DESCRIPTION
**Pull request by kaisong1990**
…ilent flow

In the PR, I put a change of reverting the using the FRT and ART (MRRT) for redeeming the AT in order to trying to resolve an IcM issue where TOU claim is gone when access a different resource within the same app. The code change itself is straight forward but need more testing across different scenarios 

Describe what this PR is trying to do.

## Type of change

- [ ] Feature work
- [ ] Bug fix
- [ ] Documentation
- [x] Engineering change
- [ ] Test
- [ ] Logging/Telemetry

## Risk

- [ ] High – Errors could cause MAJOR regression of many scenarios. (Example: new large features or high level infrastructure changes)
- [x] Medium – Errors could cause regression of 1 or more scenarios. (Example: somewhat complex bug fixes, small new features)
- [ ] Small – No issues are expected. (Example: Very small bug fixes, string changes, or configuration settings changes)

### CODE REVIEW

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 1]
```diff
@@ -46,6 +46,13 @@
```
Comment (kaisong1990) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:46]: This pull request does not update changelog.txt.

Please consider if this change would be noticeable to a partner or user and either update changelog.txt or resolve this conversation.

File: IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m [hunk 2]
```diff
@@ -1594,23 +1514,15 @@ - (void)testAcquireTokenSilent_whenFamilyRefreshTokenInCache_andClientMismatch_s
...
-        XCTAssertEqualObjects(result.accessToken.accessToken, @"new at mrrt");
-        XCTAssertEqualObjects(result.accessToken.scopes, [NSOrderedSet msidOrderedSetFromString:@"user.read tasks.read"]);
-        XCTAssertEqualObjects(result.account.accountIdentifier.homeAccountId, silentParameters.accountIdentifier.homeAccountId);
-        XCTAssertEqualObjects(result.rawIdToken, [MSIDTestIdTokenUtil idTokenWithPreferredUsername:DEFAULT_TEST_ID_TOKEN_USERNAME subject:@"sub" givenName:@"Test" familyName:@"User" name:@"Test Name" version:@"2.0" tid:DEFAULT_TEST_UTID]);
-        XCTAssertFalse(result.extendedLifeTimeToken);
-        NSURL *tenantURL = [NSURL URLWithString:@"https://login.windows.net/"DEFAULT_TEST_UTID];
-        XCTAssertEqualObjects(result.authority.url, tenantURL);
-        XCTAssertEqualObjects(result.refreshToken.refreshToken, @"new mrrt");
+        XCTAssertNil(result);
+        XCTAssertNotNil(error);
```
Comment (oldalton) [inline IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m:1518]: could you provide a bit more information why this test case was changed?

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 3]
```diff
@@ -235,126 +217,132 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
+    // When using ART or FRT, it will go through the same method below, and handle differently within the completion block
+    [self refreshToken:refreshToken
+       completionBlock:^(MSIDTokenResult * _Nullable result, NSError * _Nullable error) {
+        if (error)
+        {
+            if ([self isErrorRecoverableByUserInteraction:error])
+            {
+                NSError *msidError = nil;
+                
+                if (isAppRefreshToken)
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:305]: This code is a bit hard to follow. Would it make sense to have different error handling for app refresh token and family refresh token?

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 4]
```diff
@@ -235,126 +217,132 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
-    }
-
-    if (familyRefreshToken)
+- (void)fetchCachedRefreshTokenWithCompletionHandler:(void (^)(MSIDBaseToken<MSIDRefreshableToken> *, MSIDRefreshTokenTypes, NSError *))completionHandler
+{
+    if (!completionHandler)
     {
-        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Found Family Refresh token, using it...");
-        [self tryFRT:familyRefreshToken completionBlock:completionBlock];
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"No completionHandler when fetch local refresh token");
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:237]: nit: I think you should just return early here :) That would avoid an else condition and calling nil completion block on line 286

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 5]
```diff
@@ -158,37 +165,12 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
-                    MSID_LOG_WITH_CTX_PII(MSIDLogLevelWarning, self.requestParameters, @"Didn't find app refresh token with error: %@", MSID_PII_LOG_MASKABLE(rtError));
-                }
-                else
-                {
-                    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Found app refresh token.");
-                }
-            }
 
+            __block MSIDBaseToken<MSIDRefreshableToken> *refreshableToken = nil;
+            [self fetchCachedRefreshTokenWithCompletionHandler:^(MSIDBaseToken<MSIDRefreshableToken> *token, __unused MSIDRefreshTokenTypes tokenType, __unused NSError *error) {
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:170]: Thanks for optimizing duplicate code here :) Appreciate!

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 6]
```diff
@@ -46,6 +46,13 @@
 
 #import "MSIDAuthenticationScheme.h"
 
+typedef NS_ENUM(NSInteger, MSIDRefreshTokenTypes)
+{
+    MSIDAppRefreshTokenType = 0,
+    MSIDFamilyRefreshTokenType,
+    MSIDInvalidRefreshTokenType
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:53]: what is the invalid refresh token type? just curious.

File: IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m (hunk 2 above)
Comment (kaisong1990) [reply IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m:1518]: For now, I will just make the test pass, but will take a closed look once the main change looks good. 

To answer your question regarding this change, the test was renamed to ```testAcquireTokenSilent_whenFamilyRefreshTokenInCache_andNoAppRefresh_ClientMismatch_shouldSkip```

If we keep the original test, it will first try with ART and succeed in new flow, and will not able to shows clientMismatch_shouldSkip. So I removed the ART and make it tests against FRT directly which generates the error and no results

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 6 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m]: Removed

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 4 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:236]: Updated as suggested

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 3 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:301]: Refactored the error handling into two separated functions one for ART and the other is for FRT

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 7]
```diff
@@ -235,129 +216,145 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
+    // When using ART or FRT, it will go through the same method below, and handle differently within the completion block
+    [self redeemAccessTokenWith:refreshToken
+                completionBlock:^(MSIDTokenResult * _Nullable result, NSError * _Nullable error) {
+        if (error)
+        {
+            if ([self isErrorRecoverableByUserInteraction:error])
+            {
+                if (isAppRefreshToken)
+                {
+                    [self handleErrorResponseForAppRefreshToken:refreshToken completionBlock:completionBlock];
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:303]: It looks like there's a possibility that completionBlock will be invoked twice with the new logic:

1. Inside handleErrorResponseForAppRefreshToken:
2. At the line 310

Could you double check that completionBlock wouldn't get invoked twice?

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 8]
```diff
@@ -235,129 +216,145 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
-    if (!multiResourceRefreshToken)
+    NSError *error = nil;
+    MSIDRefreshToken *familyRefreshToken = [self familyRefreshTokenWithError:&error];
+    if (error)
     {
-        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Refresh token wasn't found, user interaction is required.");
-        
-        NSError *interactionError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractionRequired, @"User interaction is required", nil, nil, nil, self.requestParameters.correlationId, nil, YES);
-        completionBlock(nil, interactionError);
+        MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, self.requestParameters, @"Failed to retrieve Family Refresh token with error %@", MSID_PII_LOG_MASKABLE(error));
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:334]: Should we return interaction required here instead? User would then at least be able to recover by acquiring tokens interactively. With this error user would be blocked.

**Review by oldalton — CHANGES_REQUESTED** [review]
Added two more comments, otherwise changes are looking good.

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 7 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m]: Good catch, update the logic and should do a early return to stop current the code flow

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 8 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m]: Reworked to take the same interactive flow if there is no FRT available

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 9]
```diff
@@ -158,37 +164,12 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
-                }
-                else
-                {
-                    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Found app refresh token.");
-                }
-            }
 
+            __block MSIDBaseToken<MSIDRefreshableToken> *refreshableToken = nil;
+            [self fetchCachedRefreshTokenWithCompletionHandler:^(MSIDBaseToken<MSIDRefreshableToken> *token, __unused MSIDRefreshTokenTypes tokenType, __unused NSError *error) {
+                refreshableToken = token;
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:170]: nit: should we log an error here?

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 10]
```diff
@@ -235,129 +216,148 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
+    }];
+    
+}
 
-    NSError *frtCacheError = nil;
-    MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Looking for Family Refresh token...");
-    MSIDRefreshToken *familyRefreshToken = [self familyRefreshTokenWithError:&frtCacheError];
-
-    if (frtCacheError)
+- (void)fetchCachedRefreshTokenWithCompletionHandler:(void (^)(MSIDBaseToken<MSIDRefreshableToken> *, MSIDRefreshTokenTypes, NSError *))completionHandler
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:232]: nit: This seems to be a synchronous operation. Does it actually need a completion block? You can return refresh token and pass NSError as a parameter.

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 11]
```diff
@@ -235,129 +216,148 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
             }
         }
     }
+    
+    [self fetchCachedRefreshTokenWithCompletionHandler:^(MSIDBaseToken<MSIDRefreshableToken> *refreshToken, MSIDRefreshTokenTypes tokenType, NSError *error) {
+        if (error)
+        {
+            completionBlock(nil, error);
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:223]: Consider returning interaction required here instead

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 12]
```diff
@@ -235,129 +216,148 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
             }
         }
     }
+    
+    [self fetchCachedRefreshTokenWithCompletionHandler:^(MSIDBaseToken<MSIDRefreshableToken> *refreshToken, MSIDRefreshTokenTypes tokenType, NSError *error) {
+        if (error)
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:221]: I think we should check for refresh token and not error here, since when there's no refresh token in cache, there'll be no error, but we shouldn't proceed.

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 13]
```diff
@@ -235,129 +216,148 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
+                completionBlock(nil, interactionError);
+            }
+            else
+            {
+                completionBlock(nil, error);
+            }
+        }
+        else
+        {
+            completionBlock(result, nil);
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:326]: nit: consider doing this check very early to avoid one extra level if indentation

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 14]
```diff
@@ -235,129 +216,148 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
-                     {
-                         completionBlock(nil, error);
-                     }
-                 }
-                 else
-                 {
-                     completionBlock(result, nil);
-                 }
-             }];
+    return nil;
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:348]: Seems nobody invokes completion block in this case?

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 15]
```diff
@@ -235,129 +216,148 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
+    NSError *error = nil;
+    MSIDRefreshToken *familyRefreshToken = [self familyRefreshTokenWithError:&error];
+    if (error)
     {
-        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"Refresh token wasn't found, user interaction is required.");
-        
-        NSError *interactionError = MSIDCreateError(MSIDErrorDomain, MSIDErrorInteractionRequired, @"User interaction is required", nil, nil, nil, self.requestParameters.correlationId, nil, YES);
-        completionBlock(nil, interactionError);
-        return;
+        MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, self.requestParameters, @"Failed to retrieve Family Refresh token with error %@, and user interaction is required", MSID_PII_LOG_MASKABLE(error));
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:339]: Seems nobody invokes completion block in this case.

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 16]
```diff
@@ -235,129 +216,148 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
+                    {
+                        [self tryRefreshToken:familyRefreshToken
+                                    tokenType:MSIDFamilyRefreshTokenType
+                              completionBlock:completionBlock];
+                        return;
+                    }
+                }
+                
+                // Handle error case when try with Family refresh token
+                [self handleErrorResponseForFamilyRefreshToken:error];
```
Comment (oldalton) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:315]: Seems this one would get executed for app refresh token as well, which might not be correct.

**Review by oldalton — COMMENTED** [review]
Added a few more questions/comments.
Once those are resolved, I feel that the change is looking good overall, and I'd like to see more tests to verify some edge conditions after that.

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 9 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:170]: Error should be logged in ```fetchCachedRefreshTokenWithCompletionHandler``` method. No extra actions needed to be handled when there is an error in the block, and marked it as unused

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 10 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:232]: Good point, the reason I used a completion block here is due to another params called  ```MSIDRefreshTokenTypes```, where would need to distinguish the type of refresh in later code path

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 16 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:315]: Fixed

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 14 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:348]: Could you specify which completion block?

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 15 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:339]: Could you specify which completion block?

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 12 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:221]: fixed

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 11 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:223]: Updated

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 13 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:321]: Could you specify which check it is? The comment landed on a completionBlock block :)

File: IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m (hunk 2 above)
Comment (kaisong1990) [reply IdentityCore/tests/integration/ios/MSIDDefaultSilentTokenRequestTests.m:1518]: UT has been updated

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 17]
```diff
@@ -235,129 +216,143 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
     {
-        MSID_LOG_WITH_CTX_PII(MSIDLogLevelError, self.requestParameters, @"Failed to read family refresh token with error %@", MSID_PII_LOG_MASKABLE(frtCacheError));
-        completionBlock(nil, frtCacheError);
+        MSID_LOG_WITH_CTX(MSIDLogLevelInfo, self.requestParameters, @"No completionHandler when fetch local refresh token");
         return;
     }
-
-    if (familyRefreshToken)
+    
+    NSError * rtError = nil;
```
Comment (jasoncoolmax) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:241]: nit: `NSError *rtError` following our coding convention :)

Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m]: Fixed

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m [hunk 18]
```diff
@@ -235,129 +216,143 @@ - (void)executeRequestImpl:(MSIDRequestCompletionBlock)completionBlock
...
+    [self redeemAccessTokenWith:refreshToken
+                completionBlock:^(MSIDTokenResult * _Nullable result, NSError * _Nullable error) {
+        if (error)
+        {
+            if ([self isErrorRecoverableByUserInteraction:error])
+            {
+                if (isAppRefreshToken)
+                {
+                    // Handle error case when try with App refresh token
+                    MSIDBaseToken<MSIDRefreshableToken> *familyRefreshToken = [self handleErrorResponseForAppRefreshToken:refreshToken];
```
Comment (jasoncoolmax) [inline IdentityCore/src/requests/MSIDSilentTokenRequest.m:296]: Design wise, do you think it is better to move line 297-301 to `handleErrorResponseForAppRefreshToken`? as retrying PRT seems to be a part of the logic in error handler.

**Review by jasoncoolmax — APPROVED** [review]

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 13 above)
Comment (oldalton) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m:321]: You can do if (!error) (call completionBlock) as the very first thing on line 289. This would made code a bit more readable.

**Review by oldalton — APPROVED** [review]

File: IdentityCore/src/requests/MSIDSilentTokenRequest.m (hunk 18 above)
Comment (kaisong1990) [reply IdentityCore/src/requests/MSIDSilentTokenRequest.m]: Updated
